Extend pcap-bosh RFC with architecture diagram and new, simplified alternative (#1)
domdom82/community: toc/rfc/rfc-draft-pcap-bosh.md

The release was created after an initial discussion was started on the [cf-deployment](https://github.com/cloudfoundry/cf-deployment/issues/980) repository.
Add a feature to BOSH that provides a convenient network capturing mechanism for instances deployed via BOSH and easy streaming of network traffic to the user issuing the capture request.

Since then, the project has evolved into a BOSH use case and a Cloud Foundry use case. This RFC discusses a proposal to get the BOSH use case integrated with BOSH director, BOSH agent and BOSH cli.
Capture targets are identified automatically based on metadata in the BOSH Director, i.e. their deployment, instance groups and/or specific instance IDs.

The main goal of the pcap feature is to allow users (BOSH operator in this case) to easily capture network traffic across multiple VMs by specifying various parameters in PCAP CLI:
- BOSH deployment
- BOSH instance group(s)

> - BPF [Berkeley Packet Filter](https://www.tcpdump.org/manpages/pcap-filter.7.html) filter to apply

Reviewer: I wouldn't necessarily remove this section. Even the BOSH lite case needs a list of parameters to identify what to capture. Above you mentioned only "metadata, i.e. deployment, instance groups..." but I would keep the argument list intact. The only difference is that the arguments now go into BOSH CLI instead of PCAP-BOSH CLI.

Author: I retained most of this section and added the resulting capture target identification based on these parameters after.

## Proposal

The feature proposed in this RFC has two possible solutions, each of which have their benefits and drawbacks. The solution with the most community support will ultimately be implemented. Motivations, constraints and considerations are highlighted in those respective sections.

The following solutions are discussed in subsequent sections (linked below):
1. [Integrate pcap-release with BOSH](#integrate-pcap-release-with-bosh-cli-director-agent)  
   This solution focuses on integrating components that are reusable for other scenarios and ensure long-term availability, use and improvements.
2. [Add "pcap-lite", a thin front-end and merge functionality of tcpdump streams](#add-pcap-lite-a-thin-front-end-and-merge-functionality-of-tcpdump-streams)  
   This solution focuses on a convenient minimal solution for the problem statement, but presents a separate code base.

> > **Warning**

Reviewer: I don't get why a warning is needed. 
```suggestion
> **Remark**
```
?

Reviewer 2: `Remark` doesn't exist, you can have [Note, Important or Warning](https://github.com/orgs/community/discussions/16925)

Author: Maybe I can make it a Note or Important… warning is a bit strong, fair.

Author: Also thanks for the link. Apparently the new syntax is `[!NOTE]`

> [!NOTE]
> This is a note.

> While this RFC is in review, both options are presented and discussed. A final version of the RFC will highlight the chosen solution and summarise the discarded approach briefly.

### Use Case Overview

The following use cases are considered:
1. Capture data from a target deployment, filtering for:
   - instance groups, and/or
   - instance IDs

> 2. Providing detailed [Berkeley Packet Filter (BPF) filters in `libpcap` notation](https://www.tcpdump.org/manpages/pcap-filter.7.html), as used e.g. in [`tcpdump`](https://www.tcpdump.org/manpages/tcpdump.1.html) in a safe manner.

Reviewer 2: It's a [pcap-filter](https://man.archlinux.org/man/pcap-filter). It gets compiled to (e)BPF but that's none of our business so I wouldn't mention it.

Author: Fair. I left that bit from the previous version. Will consolidate with Dominiks comments.

Author: addressed now.

#### Use Cases

The cases from the [Use Case Overview](#use-case-overview) are addressed below:

| # | Use Case                                                           | Comment                                                                                                                                                                                                                                   |
|---|--------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| 1 | Capturing Data from specific instances                             | Instance groups and their instances can be determined by querying the BOSH Director with an appropriate token. When integrated into the bosh-cli, the pcap-client can access the managed token information directly.                      |

> | 2 | Providing BPF Filters                                              | BPF Filters are checked for length to avoid accidental overload or attacks. Other checks can be added centrally.                                                                                                                          |                                                                                                                                         |

Reviewer 2: Same as above: the user provides a `pcap-filter`, you get the point and I will stop mentioning it from now on.

* Comprehensive logging of requests, connection issues, congestion and successful completion.
* mTLS encrypted communication between all components, using gRPC over HTTP/2
* Authentication and authorization are handled via the existing BOSH UAA and respective BOSH CLI tokens. Accordingly, the current solution is limited to using UAA and token based authentication with RSA tokens. Other authentication mechanisms can be implemented but require additional work.

While the pcap-release is complex, it is also very flexible and versatile, with many advanced use cases that go beyond simply capturing network traffic to a client machine. While these benefits are not realized with the current state of the implementation, they are on the horizon. 

The largest short-term benefit of integrating pcap-release with BOSH is the increased user base and feedback channel that will help improve pcap-release for the mid- and long-term use and future developments.

### Add "pcap-lite", a thin front-end and merge functionality of tcpdump streams

The alternative solution is based on the orchestration of existing tools with a small amount of new code.

> Akin to the BOSH SSH feature, a BOSH pcap feature is proposed with similar but slightly different workflow. The well-known `tcpdump` tool is launched on each VM via SSH and transmits its data via the SSH channel. In order to handle multiple VMs and merge their captures, multiple SSH sessions to the respective targets are opened in parallel, multiplexed in the bosh-cli and available for writing to disk or an output stream.

Reviewer: ```suggestion
Akin to the *BOSH SSH* feature, a *BOSH pcap* feature is proposed with a similar but slightly different workflow. The well-known `tcpdump` tool is launched on each VM via SSH and transmits its data via the SSH channel. In order to handle multiple VMs and merge their captures, multiple SSH sessions to the respective targets are opened in parallel, demultiplexed in the bosh-cli and available for writing to disk or an output stream.
```

Author: @domdom82 this should still be "multiplexed", as we are combining multiple streams into one, not splitting a combined stream into different ones.

Author: Agreed to use "merged", which will not cause any confusion.

#### Architecture Overview

The architecture of the pcap-lite solution looks as follows:

![](rfc-draft-pcap-bosh/tcpdump-bosh-pcap-lite.drawio.svg "Architecture Overview for the pcap-lite Solution")

**Figure 2 - Architecture Overview for the pcap-lite Solution**

The main benefit of operator convenience over raw use of command line tools requires the following:
* Automatic determination of VM target addresses via the BOSH Director.
* A single SSH session can contain multiple streams, one for data, one for process control. The Go SSH library always opens such a "Master Mode" connection.
* Multiplexing can be done in the BOSH CLI, using a similar mechanism to pcap-release, i.e. by using the `gopacket/gopacket` libraries (and `libpcap` under the hood).

> The largest constraint is that the capturing is done used by tools that must already be available on the target VMs, e.g. `tcpdump` that is available as part of the stemcell.

Reviewer: we should mention input sanitation here, as we are no longer calling a well-defined API but forwarding user input directly to a shell command. as root.

Reviewer 2: But the whole point is that it doesn't make sense to provide those safe-guards. In this scenario you are admin anyway, why bother crafting some sort of injection attack when you can just `bosh ssh -c 'sudo rm -rf /'` right away?

Author: Also we don’t to any sanitisation in pcap-release.

The binary call via SSH should also not be `bash -c “$COMMAND”` but a well defined arguments chain. Let’s see if that works :)

Author: I mention in the pros/cons that this is not increased risk compared to having bosh ssh available.

> The largest constraint is that the capturing is done used by tools that must already be available on the target VMs, e.g. `tcpdump` that is available as part of the stemcell.

Reviewer 2: Well this "constraint" is actually a benefit of this solution. The fully-integrated approach requires us to modify the stemcell (by changing the agent) whereas this approach does not require any changes to the stemcell.

Author: addressed. Also added the bit about the power of BOSH SSH in the pros/cons list.

> Support for capturing from [Windows stemcells](https://bosh.io/docs/windows-sample-release/) must be considered further, e.g. by invoking [Npcap](https://npcap.com) instead of `tcpdump`, as suitable replacement for Windows. Please also note that [WinPcap is essentially unsupported](https://npcap.com/vs-winpcap.html). 

Reviewer: ugh. this gets uglier by the minute... but maybe VMware can provide help here since they have notable windows customers.

Author: Nobody has asked for windows yet. We can still make this less prominent.

Author: As discussed, toned down on the Windows angle.

> Support for capturing from [Windows stemcells](https://bosh.io/docs/windows-sample-release/) must be considered further, e.g. by invoking [Npcap](https://npcap.com) instead of `tcpdump`, as suitable replacement for Windows. Please also note that [WinPcap is essentially unsupported](https://npcap.com/vs-winpcap.html). 

Reviewer 2: This seems a bit unfair, we have no idea whether the other solution is capable of running on windows (and to be honest it probably won't run). So this is a point that has to be investigated either way.

Author: We might need to tone down the whole windows angle. It’s a point I wrote about but am not certain about.

Author: As discussed, the Windows angle has been relegated to a note that says "we know Windows is a thing, we haven't looked into details,  but it looks like there are approaches for both solutions."

| 5 | Output of captured data to a local file or stdout                  | The merged pcap data can be redirected to a file or configurable stream (e.g. stdout).                                                                                                                              |
| 6 | Clean shutdown and error handling                                  | `tcpdump` [accepts signals to control ongoing captures](https://www.tcpdump.org/manpages/tcpdump.1.html). This includes clean shutdown with `SIGINT`/`SIGTERM` and flushing already captured buffers via `SIGUSR2`. |
| 7 | Authorized Access                                                  | Access authorization can be enforced similar to BOSH SSH. The same SSH keys can be re-used as well.                                                                                                                 |

All of those use cases would need to be re-assessed for Windows based stemcells.

#### Summary

The "pcap-lite" solution of more advanced orchestration of streamed `tcpdump` based captures via SSH provides a lean alternative to the pcap-release.

> Some features are missing or not a extensive as with the pcap-release integration. Most notably, congestion control is missing.

Reviewer: ```suggestion
Some features are missing or not as extensive as with the pcap-release integration. Most notably, congestion control is missing.
```

For the initial problem described in this RFC, this is a "Minimal Viable Product" that provides the functionality for the support and debugging case and offers great functionality that covers a majority of use cases. 

Further developments or more advanced use cases may be hindered by its simplicity however and will need to be re-assessed as they become relevant.

### Comparison of Solutions

This section provides a comparison of both presented solutions, highlighting their benefits and drawbacks. This comparion is particularly relevant for the community's decision for either or none of the presented solutions.

#### 1. pcap-release integration into BOSH

Main benefits of the pcap-release integration approach are:
* Extended functionality through bidirectional signalling, including congestion control.
* Clear signalling of state, errors during communication and status informaiton for the user.
* Central component with enforceable limits, e.g. in terms of concurrent captures.

> * Portable Go code should support any stemcell (including Windows), for which libpcap is available.

Reviewer 2: This is a bold claim.

Author: Reduced boldness. Changed this to be about being self-contained code without the need to call external processes.

The main goal of the pcap feature is to allow users (BOSH operator in this case) to easily capture network traffic across multiple VMs by specifying various parameters on the BOSH CLI:
- BOSH deployment
- BOSH instance group(s), and optionally instance IDs
- [pcap-filter](https://www.tcpdump.org/manpages/pcap-filter.7.html) to apply to the captured traffic.

Capture targets and their IP addresses are then identified automatically based on looking up metadata from the BOSH Director using these parameters.

A single, combined stream of capture data is then provided as output and can be written to file or piped into another command.

## Problem

Distributed, and particularly networked systems are hard to debug. Software developers and operators are not necessarily the same people, in some cases not even the same companies.

Network traffic can be reasonably easily captured via tools such as `tcpdump` on local machines. In order to get a unified view of the network traffic of relevant nodes, those captures need to be retrieved, merged and analysed.

> BOSH does not have a user or operator friendly way for resolving targets, or capturing, filtering and merging network traffic.

Reviewer 2: This always throws me off, to me this list sounds like the BOSH can't provide any of the listed operations on their own, but resolving targets is just `bosh vms` isn't it? It's the combination of "resolving targets, and capturing, filtering and merging traffic", isn't it (although the double and doesn't sound great)?

Author: good catch. This sentence is confusing. I will refine it. As you say, BOSH is particularly good at identifying targets, etc. but tcpdump isn't.

The same mechanism as for BOSH instance resolution should be used, i.e. when not providing an instance group on a deployment it would take all instances, e.g.

```shell
# capture all gorouter instances
bosh pcap -d cf gorouter

# capture HAProxy instance `some-guid-here` in instance group `ha_proxy_z1`
bosh pcap -d haproxy ha_prox_z1/some-guid-here

# capture the first NATS server
bosh pcap -d cf nats/0
```

Author: @maxmoehl, please have a look at the latest state.

Reviewer 2: LGTM!

The alternative solution is based on the orchestration of existing tools with a small amount of new code.

Akin to the *BOSH SSH* feature, a *BOSH pcap* feature is proposed with a similar but slightly different workflow.
The well-known `tcpdump` tool is launched on each VM via SSH and transmits its data via the SSH channel.
In order to handle multiple VMs and merge their captures, multiple SSH sessions to the respective targets are opened in parallel, 
merged into a single stream in the bosh-cli and available for writing to disk or an output stream.

#### Architecture Overview

The architecture of the pcap-lite solution looks as follows:

![](rfc-draft-pcap-bosh/tcpdump-bosh-pcap-lite.drawio.svg "Architecture Overview for the pcap-lite Solution")

*Figure 2 - Architecture Overview for the pcap-lite Solution*

The main benefit of operator convenience over raw use of command line tools requires the following:
* Automatic determination of VM target addresses via the BOSH Director.
* A single SSH session can contain multiple streams, one for data, one for process control. The Go SSH library always opens such a "Master Mode" connection.

> * Multiplexing can be done in the BOSH CLI, using a similar mechanism to pcap-release, i.e. by using the `gopacket/gopacket` libraries (and `libpcap` under the hood).

Reviewer 2: ```suggestion
* Multiplexing can be done in the BOSH CLI, using a similar mechanism to pcap-release, i.e. by using the `github.com/gopacket/gopacket` module (and `libpcap` under the hood).
```

#### Use Cases

The cases from the [Use Case Overview](#use-case-overview) are addressed below:

| # | Use Case                                                                             | Comment                                                                                                                                                                                                             |
|---|--------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| 1 | Capturing Data from specific instances                                               | Instance groups and their instances can be determined by querying the BOSH Director with an appropriate token. The invocation of `tcpdump` can be constructed with appropriate command line arguments.              |

> | 2 | Providing [pcap-filter](https://www.tcpdump.org/manpages/pcap-filter.7.html) Filters | Filters are checked for length to avoid accidental overload or attacks, and provided to the invocation of `tcpdump` on the respective target VMs.                                                                   |                                                                                                                                         |

Reviewer 2: Are they checked? I would say no because, as discussed previously, there isn't really a reason to do so.

Author: I can write 'can be checked'. You still might want some safeguards?
But I see your point. You can do the same via BOSH SSH either way.

Author: Removed the checking bit.

Reviewer: It may be a misunderstanding that just because someone can bypass any security checks means there should be no security checks. The reason we should sanitize input is not so much to protect against hackers but to avoid accidents. Even admins make mistakes.

We should not just pass whatever user input is given straight to "sudo tcpdump -w -" as verbatim. The user should be able to provide filter input, interface, snaplength etc. but not the actual `-i`, `-s`, `-f` parameters. If we do that we have no control over what will come back, which format it may have etc. And we won't have any means to confidently merge the returned pcap stream.

Author: The idea is not to just pass on flags verbatim, but pass on the pcap-filter expression itself verbatim. interface, snaplen, etc. would become parameters to the `bosh pcap` subcommand.
